Parse data from TABLES/TABLE in DIAG-DATA-DICTIONARY-SPEC

smileprem opened the pull request:
- Parse the data from TABLES and persist them in id_lookup dict to be used elsewhere
- Added three utility functions to fetch the addresses for CAN and DoIP
- Updated .gitignore to ignore specific folders.

smileprem commented:
<img width="626" alt="image" src="https://user-images.githubusercontent.com/5611865/179293761-54000694-2a31-4bed-87c7-6116a8daf532.png">

Hi team - Could you please update the email address to send CLA?

andlaus commented:
Hi Premkumar,

Thanks for your contribution and sorry for the sluggishness; I was AFK last week, so I missed this PR until now. I promise to assign the attention it deserves to it next week :)

You can try to send the CLA to cla-mbition@mercedes-benz.com or, if this does not work either, to me personally: andreas.lauser@mercedes-benz.com .

smileprem commented:
Hi @andlaus - I sent the CLA to the email addresses mentioned above. Thanks for making this wonderful library as open source. I'll contribute back to the library as and when I can. Thank you.

andlaus commented on odxtools/table.py:
```diff
@@ -0,0 +1,113 @@
+    short_name: str
+    long_name: str
+    key: int
+    structure_ref: str
```
you should probably resolve references. the mechanism is to add a `_resolve_references(id_lookup)` method to all classes containing references and call it from the instancing class' `_respolve_references()`. in this case, you should add these method to `TableRow` and `Table` and call the latter one from  [DataDictionarySpec](https://github.com/mercedes-benz/odxtools/blob/main/odxtools/diagdatadictionaryspec.py#L75).

smileprem replied:
Sure, I'll add the recommended approach.

smileprem replied:
Hi @andlaus - I've added the functionality to resolve the references for TABLE and TABLE-ROWS. Please review. 

I'll review the templates and the test cases for tables in the next few days.

andlaus reviewed (changes requested):
Thanks a lot for your contribution, we appreciate it :)

The parsing code is mostly fine, but before it is merged, I'd like to see support for writing table objects (see e.g. the writing code for the [audience data](https://github.com/mercedes-benz/odxtools/blob/main/odxtools/pdx_stub/macros/printAudience.tpl)) and an extension of the ["Somersault" ECU](https://github.com/mercedes-benz/odxtools/blob/main/examples/somersaultecu.py). once this is done, extending [test_somersault](https://github.com/mercedes-benz/odxtools/blob/main/tests/test_somersault.py) would be nice.

smileprem commented on CONTRIBUTORS_LICENSE_AGREEMENT.md:
```diff
@@ -34,6 +34,7 @@ Germany
 (_Contributor_ and _FOSS Co_ hereinafter individually a **_“Party”_** and jointly the **_“Parties”_**)
 
 ## 1 Definitions
+
```
Sure. I'll revert the changes.

smileprem commented on README.md:
````diff
@@ -483,10 +483,15 @@ The included unit tests can be run via
 ```
 
+The static type checker can be run via
+```
````
Updated

smileprem commented on README.md:
````diff
@@ -483,10 +483,15 @@ The included unit tests can be run via
 
+The static type checker can be run via
+```
+$ python3 -m mypy --ignore-missing-imports --python-version 3.8 odxtools
````
Updated

smileprem commented on odxtools/parameters/readparameter.py:
```diff
@@ -193,10 +194,11 @@ def read_parameter_from_odx(et_element):
-                                 semantic=semantic,
                                  byte_position=byte_position,
                                  bit_position=bit_position,
+                                 semantic=semantic,
```
To reflect the same order as the class TableKeyParameter. Happy to revert if you think its not required.

```
class TableKeyParameter(Parameter):
    def __init__(self,
                 short_name,
                 table_ref=None,
                 table_snref=None,
                 table_row_snref=None,
                 table_row_ref=None,
                 id=None,
                 long_name=None,
                 byte_position=None,
                 bit_position=0,
                 semantic=None,
                 description=None):
```

andlaus replied:
no, that's okay. I was just wondering...

smileprem commented on odxtools/table.py:
```diff
@@ -0,0 +1,113 @@
+# SPDX-License-Identifier: MIT
+# Copyright (c) 2022 MBition GmbH
```
Thanks for the offer. But, I think it should be OK. Its a very small MR. :)

andlaus replied:
well, you created that file :)

Anyway, If i underdstand the legalese correctly, these copyright notices at the beginning of source files are purely informal anyway (the copyright is with whomever made the respective piece of work...)

andlaus commented on odxtools/table.py:
```diff
@@ -28,6 +29,21 @@ class TableRow:
+
+    @property
+    def structure(self) -> Optional[DopBase]:
+        """The data object property describing this parameter."""
```
```suggestion
        """The structure associated with this table."""
```

(looks like a copy-and-pasto to me...)

andlaus commented on odxtools/table.py:
```diff
@@ -51,12 +67,25 @@ def __init__(
+
+    @property
+    def key_dop(self) -> Optional[DopBase]:
+        """The data object property describing this parameter."""
```
this docstring is also wrong?

smileprem replied:
Updated the doc string

andlaus reviewed (commented):
the new reference resolution logic looks good. Thanks for the work :)

smileprem commented:
> Thanks a lot for your contribution, we appreciate it :)
> 
> The parsing code is mostly fine, but before it is merged, I'd like to see support for writing table objects (see e.g. the writing code for the [audience data](https://github.com/mercedes-benz/odxtools/blob/main/odxtools/pdx_stub/macros/printAudience.tpl)) and an extension of the ["Somersault" ECU](https://github.com/mercedes-benz/odxtools/blob/main/examples/somersaultecu.py). once this is done, extending [test_somersault](https://github.com/mercedes-benz/odxtools/blob/main/tests/test_somersault.py) would be nice.

- Added the code to write tables to pdx files
- Updated the test case in test_diag_data_dictionary_spec.py to include tables

smileprem commented on examples/somersaultecu.py:
```diff
@@ -364,6 +361,40 @@ class SID(IntEnum):
 }
 
+# tables
+somersault_tables = {
```
Please feel free to change the content of this test table

andlaus replied:
it is fine (quirky) :)

andlaus commented:
Thanks for the fixes. this looks very good now and from the coding point of view it is basically ready to be merged. There's one issue which prevents me from hitting green, though: the new generated PDX files do no longer validate:

```bash
$ ./mksomersaultpdx.py somersault.pdx
$ unzip somersault.pdx
$ xmllint --schema ~/standards/odx/BaseStandard/Schema/odx.xsd --noout somersault.odx-d
somersault.odx-d:179: element DATA-OBJECT-PROPS: Schemas validity error : Element 'DATA-OBJECT-PROPS': This element is not expected. Expected is ( SDGS ).
somersault.odx-d fails to validate
```

(This is not a big deal, most likely it is caused by a wrong order of the XML tags. if you don't have the XSD available, I can fix this for you.)

besides this, there are also two nice-to-haves before merge: it would be good if the shipped version of `somersault.pdx` would correspond to the current state of affairs (just run `mksomersault.py` in the examples directory and `git add` the result), and it would be great if the commits contained in this PR would be squashed into one (I can do this via a squash merge, but then the commit description would be of very limited use).

smileprem commented:
Thanks @andlaus. 

- Updated the order in which tables were inserted, thus passing the validation
- Added `somersault.pdx` to the codebase
- Added a README file with the instructions to validate the pdx generation logic

We generally squash the commits during merge. Please feel free to do so. Also, it would be nice to add the odx.xsd to the codebase. We could eventually add schema validation in the parsing logic too. What do you think?

andlaus commented:
Good job! Merging, thanks for your persistence!
